src/webui/api/torrentscontroller.cpp

```diff
@@ -1041,3 +1041,112 @@ void TorrentsController::categoriesAction()
 
     setResult(categories);
 }
+
+void TorrentsController::addTagsAction()
+{
+    checkParams({"hashes", "tags"});
+
+    const QStringList hashes {params()["hashes"].split('|')};
+    const QStringList tags {params()["tags"].split(',', QString::SkipEmptyParts)};
+    QJsonArray errors;
+
+    for (const QString &tag : tags) {
+        const QString tagTrimmed {tag.trimmed()};
+        applyToTorrents(hashes, [&tagTrimmed, &errors](BitTorrent::TorrentHandle *const torrent)
+        {
+            if (!torrent->addTag(tagTrimmed) && !torrent->hasTag(tagTrimmed)) {
+                errors.append(QJsonObject {
```

Comment:
There is incorrect place to handle this error. Really we are handle error that is raised due to incorrect tag, isn't it? I would check each tag before applying it. Otherwise you will get multiple error messages because of the same invalid tag.